File: Semester3/ProgammingParadigms/C5/Scala/main.scala
object app extends App {

    // 1

    def lreapeat[T](k: Int)(lxs: LazyList[T]): LazyList[T] = {
        def funHelp(n: Int, element: LazyList[T]): LazyList[T] = {
            if (n < k) {
                element match {
                    case x #:: xs => x #:: funHelp(n + 1, element)
                    case _ => LazyList.empty
                }
            }
            else {
                element match {
                    case _ #:: xs => funHelp(0, element) ++ funHelp(0, xs)
                    case _ => LazyList.empty
                }
            }
        }
        funHelp(0, lxs)
    }
 
    // 2

    val lfib = {
        def fibGen(a:Int, b: Int): Stream[Int] = {
            a #:: fibGen(b,a+b)
        }
        fibGen(0,1)
    }

    // 3

    // a

    sealed trait lBT[+A]
    case object LEmpty extends lBT[Nothing]
    case class LNode[+A](elem: A, left:()=>lBT[A], right:()=>lBT[A]) extends lBT[A]


    def lTree(n: Int): lBT[Int] = {
        LNode(n, ()=>lTree(2*n),()=>lTree(2*n+1))
    }

    // b

    def toStream[A](tree: lBT[A]) =  {
        def helper(queue: List[lBT[A]]): Stream[A] = queue match {
            case LEmpty :: tail => helper(tail)
            case LNode(value, leftSubtree, rightSubtree)::tail => value #::helper(tail++List(leftSubtree(), rightSubtree()))
            case Nil => Stream.Empty;
        }
        helper(List(tree))
    }

}

File: Semester3/ProgammingParadigms/C5/OCaml/main.ml
type 'a llist = LNil | LCons of 'a * 'a llist Lazy.t

let rec lrepeat k llist =
  let rec lrepeat2 i lazylist =
    match (i, lazylist) with
    | (0, LCons(head, tail)) -> LCons(head, lazy (lrepeat k (Lazy.force tail)))
    | (n, LCons(_, tail)) -> lrepeat2 (n - 1) (Lazy.force tail)
    | (_, LNil) -> LNil
  in lrepeat2 k llist

  (*2*)

  let rec fibgen a b =
    LCons(a, lazy (fibgen b (a + b)))
  
  let lfib = fibgen 0 1



  (*3*)

  type 'a lBT = LEmpty | LNode of 'a * (unit ->'a lBT) * (unit -> 'a lBT);;

  (*a*)

  let rec lTree n = 
      LNode(n, (function () -> lTree(2*n)), function () -> lTree(2*n+1));;

  (*b*)

  let toLazyList tree = 
    let rec helper queue = match queue with
      | [] -> LNil
      | LEmpty :: tail -> helper tail
      | LNode(value, leftSubtree, rightSubtree) :: tail ->
          LCons(value, lazy (helper (tail @ [leftSubtree(); rightSubtree()])))
    in helper [tree]